Steps4Impact/LoginV2/LoginV2ViewController.swift

```diff
@@ -0,0 +1,216 @@
+      facebookLoginButton,
+      googleLoginButton,
+      appleLoginButton
+    )
+    
+    view.addSubview(stackView) {
+      $0.top.equalTo(view.safeAreaLayoutGuide).inset(Style.Padding.p32)
+      $0.leading.trailing.bottom.equalToSuperview().inset(Style.Padding.p32)
+    }
+  }
+  
+  // MARK: - Actions
+  
+  @objc
+  private func logoLongPressed() {
+    let alert = AlertViewController()
+    alert.title = "Switch Server"
+    alert.body = "Switch to \(UserInfo.isStaging ? "Production" : "Staging")?"
+    alert.add(.okay({
+      UserInfo.isStaging.toggle()
+    }))
+    alert.add(.cancel())
+    AppController.shared.present(alert: alert, in: self, completion: nil)
+  }
+  
+  @objc
+  private func viewTermsAndConditions() {
+    let safariViewController = SFSafariViewController(
+      url: URL(string: "https://www.akfusa.org/website-private-policy")!)
+    present(safariViewController, animated: true, completion: nil)
+  }
+}
+
+// AKF-WCBackend
+extension LoginV2ViewController {
+  private func linkAKFBackend(fbid: String) {
+    AKFCausesService.createParticipant(fbid: fbid) { [weak self] (_) in
+      onMain {
+        AppController.shared.login()
+      }
+      self?.addParticipantToDefaultEventIfNeeded(fbid: fbid)
+    }
+  }
+  
+  private func addParticipantToDefaultEventIfNeeded(fbid: String) {
+    AKFCausesService.getParticipant(fbid: fbid) { (result) in
+      if let participant = Participant(json: result.response), participant.currentEvent == nil {
+        AKFCausesService.getEvents { (result) in
+          let events = result.response?.arrayValue?.compactMap { Event(json: $0) }
+
+          if let event = events?.first, let eventID = event.id {
+            AKFCausesService.joinEvent(fbid: fbid, eventID: eventID, steps: event.defaultStepCount) { _ in
+              NotificationCenter.default.post(name: .eventChanged, object: nil)
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+// Firebase
+extension LoginV2ViewController {
+  private func showFirebaseLoginErrorAlert(body: String) {
+    let alert = AlertViewController()
+    alert.title = "Error"
+    alert.body = "Error while trying to login: \(body)"
+    alert.add(.okay())
+    AppController.shared.present(alert: alert, in: self, completion: nil)
+  }
+}
+
+// Social Logins
+extension LoginV2ViewController {
+  private func linkFirebase(_ credential: AuthCredential) {
+    Auth.auth().signIn(with: credential) { [weak self] (result: AuthDataResult?, error: Error?) in
+      guard let `self` = self else { return }
+      
+      guard let error = error else {
+        if let result = result {
+          // User is signed in
+          self.linkAKFBackend(fbid: result.user.uid)
+        } else {
+          // This shouldn't happen and we don't have a way to recover except trying again.
+          self.showFirebaseLoginErrorAlert(body: "Unknown Error")
+        }
+        
+        return
+      }
+      
+      // Don't handle errors
```

Comment:
This seems confusing, can you rephrase it?  The error has been "handled" above hasn't it?

Reply:
the guard above is actually backwards above. I'll refactor it